client/state/data-layer/wpcom/login-2fa/index.js

```diff
@@ -20,71 +14,91 @@ import {
-		} ).catch( error => {
-			store.dispatch( { type: TWO_FACTOR_AUTHENTICATION_PUSH_UPDATE_NONCE, twoStepNonce: error.response.body.data.two_step_nonce } );
-			return Promise.reject( error );
-		} );
+		}
+	}, action ) );
 };
 
 /***
- * Polling the login API for the status of the push notification.
- * The polling will stop on success or when store's polling progress
- * state changes to `false`
+ * Receive data from the two factor push notification status http request
  *
- * @param {Object}   store  Global redux store
+ * @param {Function}	dispatch redux store dispatch function
+ * @returns {*} dispatch result
  */
-const doAppPushPolling = store => {
-	let retryCount = 0;
-	const retry = () => {
-		// if polling was stopped or not in progress - stop
-		if ( ! getTwoFactorPushPollInProgress( store.getState() ) ) {
-			return;
-		}
+const receivedTwoFactorPushNotificationApproved = ( { dispatch } ) =>
+	dispatch( { type: TWO_FACTOR_AUTHENTICATION_PUSH_POLL_COMPLETED } );
 
-		retryCount++;
-		setTimeout(
-			() => doAppPushRequest( store ).catch( retry ),
-			( POLL_APP_PUSH_INTERVAL_SECONDS + Math.floor( retryCount / 10 ) ) * 1000 // backoff lineary
-		);
-	};
+/***
+ * Receive error from the two factor push notification status http request
+ *
+ * @param {Object}	store  Global redux store
+ * @param {Object}	action dispathced action
+ * @param {Function}	next continue dispatch function
+ * @param {Object}	error the error object
+ */
+const receivedTwoFactorPushNotificationError = ( store, action, next, error ) => {
+	store.dispatch( { type: TWO_FACTOR_AUTHENTICATION_PUSH_UPDATE_NONCE, twoStepNonce: error.response.body.data.two_step_nonce } );
 
 	if ( getTwoFactorPushPollInProgress( store.getState() ) ) {
-		doAppPushRequest( store ).catch( retry );
+		setTimeout(
+			// eslint-disable-next-line no-use-before-define
+			() => makePushNotificationRequest( store, { type: action.type }, next ),
+			POLL_APP_PUSH_INTERVAL_SECONDS * 1000
+		);
 	}
 };
 
-const handleTwoFactorPushPoll = ( store, action, next ) => {
-	// this is deferred to allow reducer respond to TWO_FACTOR_AUTHENTICATION_PUSH_POLL_START
-	defer( () => doAppPushPolling( store ) );
-	return next( action );
+/***
+ * Dispatch a two factor push notification status request with handlers
+ */
+const makePushNotificationRequest = dispatchRequest(
+	requestTwoFactorPushNotificationStatus,
+	receivedTwoFactorPushNotificationApproved,
+	receivedTwoFactorPushNotificationError
+);
+
+/***
+ * Starts polling for push notification status
+ *
+ * @param {Object}	store  Global redux store
+ * @param {Object}	action dispatched action
+ * @param {Function}	next continue dispatch function
+ * @param {Object}	error the error object
+ * @returns {*} whatever requestTwoFactorPushNotificationStatus returns, which is undefined
+ */
+const startPolling = ( store, action, next ) => {
+	next( action ); // allow the reducer change status of polling to inProgress
```

Comment:
note that in #14025 we're making a breaking change to the data layer API where we ~won't have to~ shouldn't call `next( action )` here. only marking this down so we don't lose it in the merges